Redesign & Refac Fund, Campaign & Pledge Screen

GlenDsza opened the pull request:
**What kind of change does this PR introduce?**

feature, redesign, refactoring

**Issue Number:**

Fixes #1927 

**Did you add tests for your changes?**

Yes

**Snapshots/Videos:**

Fund Screen:
![image](https://github.com/PalisadoesFoundation/talawa-admin/assets/109027247/56b59957-c2e6-4cb9-9b47-24d1c2571364)

Campaign Screen:
![image](https://github.com/PalisadoesFoundation/talawa-admin/assets/109027247/f823ce43-af27-4df2-9c32-4ec846d4c0a9)

Pledge Screen:
![image](https://github.com/PalisadoesFoundation/talawa-admin/assets/109027247/cae36a64-8e92-495a-9d25-aa2558d7d032)

Video:

https://github.com/PalisadoesFoundation/talawa-admin/assets/109027247/46c526ad-d9f4-4daa-9542-29c5d5cc3984

**Summary**

- Redesigned & Refactored (modularized) Fund Screen and its respective create/update/delete modals (popups).
- Redesigned & Refactored (modularized) Campaign Screen and its respective create/update/delete modals (popups).
- Added Breadcrumb for efficient navigation back and forth.
- Added search & sort functionality in Fund/Campaign Screen
- Added campaign progress indicator in Pledge Screen (For now static).
- Added testcases for above changes

**Does this PR introduce a breaking change?**

No

**Other information**

- The progress indicator in Pledge Screen is for now static, as currently we miss the screens/ability for user to pledge and pay for same. In the following PR this feature would be dynamically integrated.

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

Yes 


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced sorting functionality to the fund management UI.
  - Implemented `DataGrid` for displaying fund information.
  - Added `FundModal` and `FundDeleteModal` for improved fund management.

- **Improvements**
  - Refined text labels and messages across multiple languages for better clarity and consistency.
  - Enhanced campaign overview display with breadcrumb navigation and additional details.
  - Updated styles for better user interaction and visual consistency.

- **Bug Fixes**
  - Corrected translations and terminology in various languages for accuracy.

- **Refactor**
  - Restructured components and adjusted state management in `OrganizationFunds.tsx` for improved code maintainability and readability.
  - Consolidated fund modals into a single `FundModal` component.
  - Improved state management and type safety in `FundCampaignPledge.tsx`.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

GlenDsza commented on src/screens/OrganizationFundCampaign/CampaignDeleteModal.tsx:
```diff
@@ -1,29 +1,53 @@
 import React from 'react';
 import { Button, Modal } from 'react-bootstrap';
 import styles from './OrganizationFundCampaign.module.css';
+import { useMutation } from '@apollo/client';
+import { DELETE_CAMPAIGN_MUTATION } from 'GraphQl/Mutations/CampaignMutation';
```
Invalid Suggestion: Folder containing mutation queries is named as `GraphQl`, hence the following path is used

GlenDsza commented on src/screens/OrganizationFundCampaign/CampaignDeleteModal.tsx:
```diff
@@ -1,29 +1,53 @@
+
+  const [deleteCampaign] = useMutation(DELETE_CAMPAIGN_MUTATION);
+
+  const deleteCampaignHandler = async (): Promise<void> => {
+    try {
+      await deleteCampaign({
+        variables: {
+          id: campaign?._id,
+        },
+      });
+      toast.success(t('deletedCampaign'));
+      refetchCampaign();
+      hide();
+    } catch (error: unknown) {
+      toast.error((error as Error).message);
+    }
```
The toast is displayed on delete Action, hence giving additional context of origin of error is not required. And error.message sent from server is enough for context

GlenDsza commented on src/screens/OrganizationFunds/OrganizationFunds.tsx:
```diff
@@ -1,77 +1,89 @@
+  const closeModal = (modal: Modal): void =>
+    setModalState((prevState) => ({ ...prevState, [modal]: false }));
+
+  const handleOpenModal = useCallback(
+    (fund: InterfaceFundInfo | null, mode: 'edit' | 'create'): void => {
+      setFund(fund);
+      setFundModalMode(mode);
+      openModal(Modal.SAME);
+    },
+    [openModal],
+  );
```
Modal should be opened only on certain actions, triggering modal opening on keyboard events doesn't make sense

GlenDsza commented on src/screens/OrganizationFunds/OrganizationFunds.tsx:
```diff
@@ -9,69 +1,89 @@
-  const [fundUpdateModalIsOpen, setFundUpdateModalIsOpen] =
-    useState<boolean>(false);
 
-  const [taxDeductible, setTaxDeductible] = useState<boolean>(true);
-  const [isArchived, setIsArchived] = useState<boolean>(false);
-  const [isDefault, setIsDefault] = useState<boolean>(false);
+  if (!orgId) {
+    return <Navigate to={'/'} replace />;
+  }
```
Adding fallback i.e. default value for orgId, not possible as its dynamic representing organizationId

GlenDsza commented on src/screens/OrganizationFunds/OrganizationFunds.tsx:
```diff
@@ -9,69 +1,89 @@
+    'createdAt_DESC',
+  );
+
+  const [modalState, setModalState] = useState<{ [key in Modal]: boolean }>({
+    [Modal.SAME]: false,
+    [Modal.DELETE]: false,
   });
+  const [fundModalMode, setFundModalMode] = useState<'edit' | 'create'>(
+    'create',
+  );
```
Renaming Enum Modal to ModalState doesn't make any change, and is not required

GlenDsza replied:
Anyways done

GlenDsza commented on src/screens/OrganizationFunds/OrganizationFunds.tsx:
```diff
@@ -238,187 +137,250 @@
+      headerClassName: `${styles.tableHeader}`,
+      renderCell: (params: GridCellParams) => {
+        return (
+          <div
+            className="d-flex justify-content-center fw-bold"
+            data-testid="fundName"
+            onClick={() => handleClick(params.row._id as string)}
+          >
```
Adding keyboard events is not good, as user may be directed to other screen if unaware of keyboard events

GlenDsza commented on src/screens/OrganizationFunds/OrganizationFunds.tsx:
```diff
@@ -238,187 +137,250 @@
+      headerClassName: `${styles.tableHeader}`,
+      renderCell: (params: GridCellParams) => {
+        return (
+          <Button
+            variant="outline-success"
+            size="sm"
+            className="rounded"
+            onClick={() => handleClick(params.row._id as string)}
+            data-testid="viewBtn"
+          >
+            <i className="fa fa-eye me-1" />
+            {t('viewCampaigns')}
+          </Button>
+        );
```
Adding keyboard events is not good, as user may be directed to other screen if unaware of keyboard events

GlenDsza commented on src/screens/OrganizationFunds/OrganizationFunds.module.css:
```diff
@@ -49,12 +38,33 @@
 .titlemodal {
+  color: #707070;
   font-weight: 600;
-  font-size: 25px;
-  margin-top: 1rem;
+  font-size: 32px;
   width: 65%;
+  margin-bottom: 0px;
```
CSS styling could be handled in other separate PR

GlenDsza commented on src/screens/OrganizationFunds/OrganizationFunds.tsx:
```diff
@@ -9,69 +1,89 @@
+import FundModal from './FundModal';
 import styles from './OrganizationFunds.module.css';
-import {
-  Paper,
-  Table,
-  TableBody,
-  TableCell,
-  TableContainer,
-  TableHead,
-  TableRow,
-  styled,
-  tableCellClasses,
-} from '@mui/material';
+import { Stack } from '@mui/material';
 import dayjs from 'dayjs';
 import { FUND_LIST } from 'GraphQl/Queries/fundQueries';
+import {
+  DataGrid,
+  type GridCellParams,
+  type GridColDef,
+} from '@mui/x-data-grid';
```
Done

GlenDsza commented on src/screens/OrganizationFunds/OrganizationFunds.tsx:
```diff
@@ -9,69 +1,89 @@
-}));
+};
+
+enum Modal {
+  SAME = 'same',
+  DELETE = 'delete',
+}
```
Done

GlenDsza commented on src/screens/OrganizationFunds/OrganizationFunds.tsx:
```diff
@@ -238,187 +137,250 @@
     );
   }
 
+  const columns: GridColDef[] = [
+    {
+      field: 'id',
+      headerName: 'Sr. No.',
+      flex: 1,
+      minWidth: 100,
+      align: 'center',
+      headerAlign: 'center',
+      headerClassName: `${styles.tableHeader}`,
+      sortable: false,
+      renderCell: (params: GridCellParams) => {
+        return <div>{params.row.id}</div>;
+      },
+    },
```
Adding keyboard events such as onKey will lead users being redirected unwantedly if unaware of such events

GlenDsza commented on src/screens/OrganizationFunds/OrganizationFunds.tsx:
```diff
@@ -9,69 +1,89 @@
-    fontSize: 14,
+  '&.MuiDataGrid-root .MuiDataGrid-columnHeader:focus-within': {
+    outline: 'none',
   },
-}));
-
-const StyledTableRow = styled(TableRow)(() => ({
-  '&:last-child td, &:last-child th': {
-    border: 0,
+  '& .MuiDataGrid-row:hover': {
+    backgroundColor: 'transparent',
+  },
+  '& .MuiDataGrid-row.Mui-hovered': {
+    backgroundColor: 'transparent',
+  },
+  '& .MuiDataGrid-root': {
+    borderRadius: '0.5rem',
+  },
+  '& .MuiDataGrid-main': {
+    borderRadius: '0.5rem',
   },
-}));
+};
```
Not required

GlenDsza commented on src/screens/OrganizationFunds/OrganizationFunds.tsx:
```diff
@@ -238,187 +137,250 @@
+      headerClassName: `${styles.tableHeader}`,
+      renderCell: (params: GridCellParams) => {
+        return (
+          <>
+            <Button
+              variant="success"
+              size="sm"
+              className="me-2 rounded"
+              data-testid="editFundBtn"
+              onClick={() =>
+                handleOpenModal(params.row as InterfaceFundInfo, 'edit')
+              }
+            >
+              <i className="fa fa-edit" />
+            </Button>
+            <Button
+              size="sm"
+              variant="danger"
+              className="rounded"
+              data-testid="deleteFundBtn"
+              onClick={() => handleDeleteClick(params.row as InterfaceFundInfo)}
+            >
+              <i className="fa fa-trash" />
+            </Button>
+          </>
+        );
+      },
```
Adding keyboard events such as onKey will lead users being redirected unwantedly if unaware of such events

GlenDsza commented on src/screens/OrganizationFunds/FundDeleteModal.tsx:
```diff
@@ -0,0 +1,79 @@
+import React from 'react';
+import { useTranslation } from 'react-i18next';
```
Folder containing mutation queries is named GraphQl not graphql

GlenDsza commented on src/screens/OrganizationFunds/FundDeleteModal.tsx:
```diff
@@ -0,0 +1,79 @@
+
+  const [deleteFund] = useMutation(REMOVE_FUND_MUTATION);
+
+  const deleteFundHandler = async (): Promise<void> => {
+    try {
+      await deleteFund({
+        variables: {
+          id: fund?._id,
+        },
+      });
+      refetchFunds();
+      hide();
+      toast.success(t('fundDeleted'));
+    } catch (error: unknown) {
+      toast.error((error as Error).message);
+    }
```
The toast is displayed on delete Action, hence giving additional context of origin of error is not required. And error.message sent from server is enough for context

GlenDsza commented on src/screens/OrganizationFundCampaign/OrganizationFundCampagins.tsx:
```diff
@@ -60,230 +52,282 @@ const orgFundCampaign = (): JSX.Element => {
+  const [searchTerm, setSearchTerm] = useState('');
+  const [sortBy, setSortBy] = useState<string | null>(null);
+
+  const [modalState, setModalState] = useState<{ [key in Modal]: boolean }>({
+    [Modal.SAME]: false,
+    [Modal.DELETE]: false,
   });
+  const [campaignModalMode, setCampaignModalMode] = useState<'edit' | 'create'>(
+    'create',
+  );
```
Done

GlenDsza commented on src/screens/OrganizationFundCampaign/OrganizationFundCampagins.tsx:
```diff
@@ -60,230 +52,282 @@
+    setModalState((prevState) => ({ ...prevState, [modal]: false }));
+
+  const handleOpenModal = useCallback(
+    (campaign: InterfaceCampaignInfo | null, mode: 'edit' | 'create'): void => {
+      setCampaign(campaign);
+      setCampaignModalMode(mode);
+      openModal(Modal.SAME);
+    },
+    [openModal],
+  );
```
keyboard events not required

GlenDsza commented on src/screens/OrganizationFundCampaign/OrganizationFundCampagins.tsx:
```diff
@@ -9,49 +1,49 @@
-    fontSize: 14,
+  '&.MuiDataGrid-root .MuiDataGrid-columnHeader:focus-within': {
+    outline: 'none',
   },
-}));
-
-const StyledTableRow = styled(TableRow)(() => ({
-  '&:last-child td, &:last-child th': {
-    border: 0,
+  '& .MuiDataGrid-row:hover': {
+    backgroundColor: 'transparent',
+  },
+  '& .MuiDataGrid-row.Mui-hovered': {
+    backgroundColor: 'transparent',
+  },
+  '& .MuiDataGrid-root': {
+    borderRadius: '0.5rem',
   },
-}));
+  '& .MuiDataGrid-main': {
+    borderRadius: '0.5rem',
+  },
+};
```
not required

GlenDsza commented on src/screens/OrganizationFundCampaign/OrganizationFundCampagins.tsx:
```diff
@@ -60,230 +52,282 @@
+      sortable: false,
+      renderCell: (params: GridCellParams) => {
+        return (
+          <Button
+            variant="outline-success"
+            size="sm"
+            className="rounded"
+            data-testid="viewBtn"
+            onClick={() => handleClick(params.row.campaign._id as string)}
+          >
+            <i className="fa fa-eye me-1" />
+            {t('viewPledges')}
+          </Button>
+        );
```
keyboard events not required

GlenDsza commented on src/screens/OrganizationFundCampaign/OrganizationFundCampagins.tsx:
```diff
@@ -60,230 +52,282 @@
+      sortable: false,
+      renderCell: (params: GridCellParams) => {
+        return (
+          <div
+            className="d-flex justify-content-center fw-bold"
+            data-testid="campaignName"
+            onClick={() => handleClick(params.row.campaign._id as string)}
+          >
```
keyboard events not required

GlenDsza commented on src/screens/OrganizationFundCampaign/OrganizationFundCampagins.tsx:
```diff
@@ -60,230 +52,282 @@
-    useState<boolean>(false);
-  const [campaignDeleteModalIsOpen, setcampaignDeleteModalIsOpen] =
-    useState<boolean>(false);
+  const { fundId, orgId } = useParams();
+
+  if (!fundId || !orgId) {
+    return <Navigate to={'/'} />;
+  }
```
No need of adding fallback, as we are handling scenario where they could be undefined

GlenDsza commented on src/screens/OrganizationFundCampaign/OrganizationFundCampagins.tsx:
```diff
@@ -9,49 +1,49 @@
-  InterfaceCreateCampaign,
   InterfaceQueryOrganizationFundCampaigns,
 } from 'utils/interfaces';
-import CampaignCreateModal from './CampaignCreateModal';
+import CampaignModal from './CampaignModal';
 import CampaignDeleteModal from './CampaignDeleteModal';
-import CampaignUpdateModal from './CampaignUpdateModal';
 import styles from './OrganizationFundCampaign.module.css';
-import {
-  Paper,
-  Table,
-  TableBody,
-  TableCell,
-  TableContainer,
-  TableHead,
-  TableRow,
-  styled,
-  tableCellClasses,
-} from '@mui/material';
+import type { GridCellParams, GridColDef } from '@mui/x-data-grid';
+import { DataGrid } from '@mui/x-data-grid';
+import { currencySymbols } from 'utils/currency';
+import { Stack, Typography, Breadcrumbs, Link } from '@mui/material';
```
Done

GlenDsza commented on src/screens/OrganizationFundCampaign/OrganizationFundCampagins.tsx:
```diff
@@ -9,49 +1,49 @@
+  },
+};
+
+enum Modal {
+  SAME = 'same',
+  DELETE = 'delete',
+}
```
Done

GlenDsza commented on src/screens/OrganizationFundCampaign/OrganizationFundCampagins.tsx:
```diff
@@ -60,230 +52,282 @@
+      renderCell: (params: GridCellParams) => {
+        return (
+          <>
+            <Button
+              variant="success"
+              size="sm"
+              className="me-2 rounded"
+              data-testid="editCampaignBtn"
+              onClick={() =>
+                handleOpenModal(
+                  params.row.campaign as InterfaceCampaignInfo,
+                  'edit',
+                )
+              }
+            >
+              <i className="fa fa-edit" />
+            </Button>
+            <Button
+              size="sm"
+              variant="danger"
+              className="rounded"
+              data-testid="deleteCampaignBtn"
+              onClick={() =>
+                handleDeleteClick(params.row.campaign as InterfaceCampaignInfo)
+              }
+            >
+              <i className="fa fa-trash" />
+            </Button>
+          </>
```
keyboard events, not required

GlenDsza commented on src/screens/FundCampaignPledge/FundCampaignPledge.module.css:
```diff
@@ -125,3 +126,76 @@
   outline: none;
 }
+
+.overviewContainer {
+  display: flex;
+  gap: 7rem;
+  width: 100%;
+  justify-content: space-between;
+  margin: 1.5rem 0 0 0;
+  padding: 1.25rem 2rem;
+  background-color: rgba(255, 255, 255, 0.591);
+
+  box-shadow: rgba(0, 0, 0, 0.16) 0px 1px 4px;
+  border-radius: 0.5rem;
+}
```
Done

GlenDsza commented on src/screens/OrganizationFunds/OrganizationFunds.tsx:
```diff
@@ -238,187 +139,250 @@ const organizationFunds = (): JSX.Element => {
+      headerClassName: `${styles.tableHeader}`,
+      renderCell: (params: GridCellParams) => {
+        return (
+          <div
+            className="d-flex justify-content-center fw-bold"
+            data-testid="fundName"
+            onClick={() => handleClick(params.row._id as string)}
+          >
```
Not required

GlenDsza commented on src/screens/OrganizationFunds/OrganizationFunds.tsx:
```diff
@@ -4,74 +1,91 @@
-    fontSize: 14,
+  '&.MuiDataGrid-root .MuiDataGrid-columnHeader:focus-within': {
+    outline: 'none',
   },
-}));
-
-const StyledTableRow = styled(TableRow)(() => ({
-  '&:last-child td, &:last-child th': {
-    border: 0,
+  '& .MuiDataGrid-row:hover': {
+    backgroundColor: 'transparent',
+  },
+  '& .MuiDataGrid-row.Mui-hovered': {
+    backgroundColor: 'transparent',
   },
-}));
+  '& .MuiDataGrid-root': {
+    borderRadius: '0.5rem',
+  },
+  '& .MuiDataGrid-main': {
+    borderRadius: '0.5rem',
+  },
+};
```
Not required

GlenDsza commented on src/screens/OrganizationFunds/OrganizationFunds.tsx:
```diff
@@ -238,187 +139,250 @@
+      headerClassName: `${styles.tableHeader}`,
+      renderCell: (params: GridCellParams) => {
+        return (
+          <>
+            <Button
+              variant="success"
+              size="sm"
+              className="me-2 rounded"
+              data-testid="editFundBtn"
+              onClick={() =>
+                handleOpenModal(params.row as InterfaceFundInfo, 'edit')
+              }
+            >
+              <i className="fa fa-edit" />
+            </Button>
+            <Button
+              size="sm"
+              variant="danger"
+              className="rounded"
+              data-testid="deleteFundBtn"
+              onClick={() => handleDeleteClick(params.row as InterfaceFundInfo)}
+            >
+              <i className="fa fa-trash" />
+            </Button>
+          </>
+        );
+      },
```
Not required

GlenDsza commented on src/screens/OrganizationFunds/OrganizationFunds.tsx:
```diff
@@ -238,187 +139,250 @@
-              <div>
-                <Button
-                  variant="success"
-                  onClick={showCreateModal}
-                  data-testid="createFundBtn"
-                  className={styles.createFundBtn}
+                <Sort className={'me-1'} />
+                {tCommon('sort')}
+              </Dropdown.Toggle>
+              <Dropdown.Menu>
+                <Dropdown.Item
+                  onClick={() => setSortBy('createdAt_DESC')}
+                  data-testid="createdAt_DESC"
                 >
-                  <i className={'fa fa-plus me-2'} />
-                  {t('createFund')}
-                </Button>
-              </div>
-            </div>
+                  {t('createdLatest')}
+                </Dropdown.Item>
+                <Dropdown.Item
+                  onClick={() => setSortBy('createdAt_ASC')}
+                  data-testid="createdAt_ASC"
+                >
+                  {t('createdEarliest')}
+                </Dropdown.Item>
+              </Dropdown.Menu>
+            </Dropdown>
+          </div>
+          <div>
+            <Button
+              variant="success"
+              onClick={() => handleOpenModal(null, 'create')}
+              style={{ marginTop: '11px' }}
+              data-testid="createFundBtn"
+            >
+              <i className={'fa fa-plus me-2'} />
+              {t('createFund')}
+            </Button>
```
not required

GlenDsza commented on src/screens/OrganizationFunds/OrganizationFunds.tsx:
```diff
@@ -4,74 +1,91 @@
+  },
+};
+
+enum ModalState {
+  SAME = 'same',
+  DELETE = 'delete',
+}
```
not required

GlenDsza commented on src/screens/OrganizationFundCampaign/OrganizationFundCampagins.tsx:
```diff
@@ -1,57 +1,52 @@
+  },
+};
+
+enum ModalState {
+  SAME = 'same',
+  DELETE = 'delete',
+}
```
not required

GlenDsza commented:
@tasneemkoushar Added code to view additional volunteers in a scroll-enabled popup. Plz review

https://github.com/PalisadoesFoundation/talawa-admin/assets/109027247/2ddec728-4382-49ea-9b77-474b985bccb6

GlenDsza commented:
> 1. The term `volunteers` is inappropriate as they are not volunteering to manually assist, but are making a financial pledge to a fund or campaign.
> 2. They are `pledgers`. Please make the update to the appropriate pledge screens.
> 3. Also, the amount raised isn't valid until funds are paid. The term `Raised` should be `Pledged`. We don't have a payment acceptance plugin yet.

@palisadoes Agree we don't have plugin for fund acceptance, but for now have kept it as constant and will add mock functionality of payments in user screen for pledges in [Link](https://github.com/PalisadoesFoundation/talawa-admin/issues/2070). So accordingly UI could be designed. Once we integrate payment plugins we can actually add a payment interface before paying.

palisadoes commented:
- In that case shouldn't we have two chart lines on the dashboard? One for Raised and the other for Donated? 
- The Donated line could be blank for now so that we remember to fix it.

GlenDsza commented:
@palisadoes adding 2 progress bar wont look good and make dashboard over crowded with similar components, would propose following options:
1. add a dropdown with down arrow on raised/pledged text in order to switch between raised and pledged. 
2. Show a group radio button to switch from pledged to raised 

![image](https://github.com/PalisadoesFoundation/talawa-admin/assets/109027247/884ec3a4-931a-446c-a8ea-3c1aa50de9dc)

palisadoes commented:
Yes, that would be better

tasneemkoushar commented:
@GlenDsza can you please share a clip of the new screen after making all the above changes mentioned in the comments by @palisadoes. 

1. Can you change the name from Volunteer to pledger.
2. Also in the below image what will be the view if we have around 5-10 pledgers. In the clip, can you try to add more than 5 pledgers and show it.
![image](https://github.com/PalisadoesFoundation/talawa-admin/assets/58592306/d48e2137-9ef7-4966-a01b-164800c191a5)

GlenDsza commented on src/screens/FundCampaignPledge/PledgeModal.tsx:
```diff
@@ -190,31 +190,31 @@ const PledgeModal: React.FC<InterfacePledgeModal> = ({
               multiple
               className={`${styles.noOutline} w-100`}
               limitTags={2}
-              data-testid="volunteerSelect"
-              options={volunteers}
+              data-testid="pledgerSelect"
+              options={pledgers}
               value={pledgeUsers}
               isOptionEqualToValue={(option, value) => option._id === value._id}
               filterSelectedOptions={true}
-              getOptionLabel={(member: InterfacePledgeVolunteer): string =>
+              getOptionLabel={(member: InterfacePledger): string =>
                 `${member.firstName} ${member.lastName}`
               }
               onChange={
                 /*istanbul ignore next*/
-                (_, newVolunteers): void => {
+                (_, newPledgers): void => {
                   setFormState({
                     ...formState,
-                    pledgeUsers: newVolunteers,
+                    pledgeUsers: newPledgers,
                   });
                 }
               }
               renderInput={(params) => (
-                <TextField {...params} label="Volunteers" />
+                <TextField {...params} label="Pledgers" />
```
pledgeUsers will always be an array either containing list of users associated with pledge or empty array, so indirectly  this scenario is handled.

GlenDsza commented on src/screens/FundCampaignPledge/PledgeModal.tsx:
```diff
@@ -78,7 +78,7 @@ const PledgeModal: React.FC<InterfacePledgeModal> = ({
     if (memberData) {
       /*istanbul ignore next*/
-      setVolunteers(memberData.organizations[0].members);
+      setPledgers(memberData.organizations[0].members);
```
before setting state variable I am already checking for existence of `memberData` and memberData.organizations can never be null as each member is linked to at least a default organization.

GlenDsza commented:
- The scenario in which we have lots of pledgers, the popup is a scroll enabled container containing pledgers
- Changed volunteers to pledgers
- Added support to view progress bar for both pledged amount and raised amount

https://github.com/PalisadoesFoundation/talawa-admin/assets/109027247/ea81c213-c876-4523-8584-3a5a493d6c21

Nitya-Pasrija reviewed (commented):
Looks good @GlenDsza 

Quick question: What happens when the goal is reached? Do we start mapping the excess?? I hope the site doesn't break cuz of slider excess

GlenDsza commented:
> Looks good @GlenDsza
> 
> Quick question: What happens when the goal is reached? Do we start mapping the excess?? I hope the site doesn't break cuz of slider excess

@Nitya-Pasrija progress bar will show completely filled if amount crosses set goal.

![image](https://github.com/PalisadoesFoundation/talawa-admin/assets/109027247/c0dc55d6-b51f-4ff0-9782-d1ceece56474)

In above image campaign goal was $3000 and the amount was $10000

palisadoes commented:
@tasneemkoushar PTAL

tasneemkoushar commented:
@GlenDsza LGTM feel free to merge
